pandas/src/util.pxd

```diff
@@ -59,6 +59,10 @@ cdef inline object get_value_at(ndarray arr, object loc):
 cdef inline set_value_at(ndarray arr, object loc, object value):
     cdef:
         Py_ssize_t i, sz
+
+    if not cnp.PyArray_ISWRITEABLE(arr):
```

Comment:
It looks like this may actually be slowing things down slightly. I will see if I can make up that time somewhere else. If we cache at a per-array level then what should be do if someone changes the flags on the underlying array between setitem calls?